src/sentry/tasks/deletion/hybrid_cloud.py

```diff
@@ -306,3 +292,83 @@ def _process_tombstone_reconciliation(
         )
 
     return has_more
+
+
+def _get_model_ids_for_tombstone_cascade(
+    tombstone_cls: type[TombstoneBase],
+    model: type[Model],
+    field: HybridCloudForeignKey,
+    watermark_target,
+    watermark_batch: WatermarkBatch,
+) -> tuple[list[int], datetime.datetime]:
+    """
+    Queries the database or databases if spanning multiple), and returns
+     a tuple with a list of row IDs to delete, and the oldest
+     tombstone timestamp for the batch.
+
+    :param tombstone_cls: Either a RegionTombstone or ControlTombstone, depending on
+     which silo the tombstone process is running.
+    :param model: The model with a HybridCloudForeignKey.
+    :param field: The HybridCloudForeignKey field from the model to process.
+    :return:
+    """
+
+    to_delete_ids = []
+    oldest_seen = timezone.now()
+    tombstone_and_model_in_same_db = router.db_for_read(model) == router.db_for_read(tombstone_cls)
+
+    if tombstone_and_model_in_same_db:
+        with connections[router.db_for_read(model)].cursor() as conn:
+            conn.execute(
+                f"""
+                    SELECT r.id, t.created_at
+                    FROM {model._meta.db_table} r
+                    JOIN {tombstone_cls._meta.db_table} t
+                        ON t.table_name = %(table_name)s AND t.object_identifier = r.{field.name}
+                    WHERE {watermark_target}.id > %(low)s AND {watermark_target}.id <= %(up)s
+                """,
+                {
+                    "table_name": field.foreign_table_name,
+                    "low": watermark_batch.low,
+                    "up": watermark_batch.up,
+                },
+            )
+
+            for (row_id, tomb_created) in conn.fetchall():
+                to_delete_ids.append(row_id)
+                oldest_seen = min(oldest_seen, tomb_created)
+
+            return to_delete_ids, oldest_seen
+
+    return get_ids_for_tombstone_cascade_cross_db(
+        tombstone_cls=tombstone_cls, model=model, field=field, watermark_batch=watermark_batch
+    )
+
+
+def get_ids_for_tombstone_cascade_cross_db(
+    tombstone_cls: type[TombstoneBase],
+    model: type[Model],
+    field: HybridCloudForeignKey,
+    watermark_batch: WatermarkBatch,
+) -> tuple[list[int], datetime.datetime]:
+    if not options.get("hybrid_cloud.allow_cross_db_tombstones"):
+        raise Exception("Cannot process tombstones due to model living in separate database.")
+
+    oldest_seen = timezone.now()
+    # Because tombstones can span multiple databases, we can't always rely on
+    # the join code above. Instead, we have to manually query the tombstone IDs
+    # first, then query the model for matching rows.
+    tombstone_entries = tombstone_cls.objects.filter(
+        id__lte=watermark_batch.up, id__gt=watermark_batch.low
+    ).values_list("object_identifier", "created_at")
```

Comment:
Should this have the `table_name` condition that the joined version has?

Reply:
Oops, yes it should 🙈